Composer/packages/client/src/components/EditableField.tsx

```diff
@@ -192,13 +197,29 @@ const EditableField: React.FC<EditableFieldProps> = (props) => {
       e.stopPropagation();
     }
     const enterOnField = e.key === 'Enter' && hasFocus;
-    if (enterOnField && !multiline) {
+    if (enterOnField && !multiline && e.shiftKey) {
+      e.stopPropagation();
+      if (onNewLine) {
+        onNewLine();
+        return;
+      }
       setMultiline(true);
+      updateField('value', e.target.value + '\n');
+      // wait for the textarea to be rendered and then restore focus and selection
+      setTimeout(() => {
+        const len = fieldRef.current?.value?.length;
+        fieldRef.current?.focus();
+        if (len) {
+          fieldRef.current?.setSelectionRange(len, len);
+        }
+      }, 100);
```

Comment:
is there a better way to do this then using `setTimeout()`? I noticed this same trick being used in `CellFocusZone`

Reply:
The `useLayoutEffect` hook would fit better here. There are two problems with it:
- It is not portable. By this I mean it can't be used in-place in the callback and requires adding more state to the component for a one-time thing
- The hook doesn't account for multiple re-renders happened in a row or with a small delay, so it won't cover the same case `setTimeout` covers here.

The current approach has an advantage in terms of that we have a control on a delay when it makes sense to account for re-renders. As an example 100ms should be a good value to return focus back if nothing is focused. We clearly can't do move focus back after one second because it might be confusing for end users as the focus can be moved anywhere else.

I was thinking on a custom hook implementation for both scenarios where we use `setTimeout` in the PR. The hook  may need to maintain the same 100ms delay though because of the reasons mentioned above.

Reply:
> It is not portable. By this I mean it can't be used in-place in the callback and requires adding more state to the component for a one-time thing

I see what you mean here, you can't exactly call the logic inside of `useLayoutEffect` explicitly, so we would have to maintain some sort of state to trigger this which could get ugly.

> The hook doesn't account for multiple re-renders happened in a row or with a small delay, so it won't cover the same case setTimeout covers here.

We could potentially mitigate this by using a debounced function that eventually focuses the element.

After reading your response and thinking about it I think I'm fine with using this approach for now as it is only being used in these two places. If it starts to become something we need to use more commonly throughout the app then we can revisit the implementation.

Lastly, I like the idea of extracting this logic out into a hook that can be used by both components.

Reply:
Sounds good. I am going to investigate the hook approach, I don't see any other place we use it at the moment though

Reply:
Updated. So the hook does the following:
- place a callback to the macro tasks queue
- if re-render happen (react uses MessageChannel which guarantees it will be executed before setTimeout), pull out the task from queue and schedule another one
- since at this point it executes macro tasks, the task will be delayed which should give React scheduler time to schedule anything else if it wants to

With Fiber there may be different scheduler task priorities, so called out that this hook works only for immediate tasks

Reply:
Awesome!